src/state/block_tree.rs

```diff
@@ -0,0 +1,565 @@
+        Ok(self.0.block_data_len(block)?)
+    }
+
+    pub(crate) fn block_data(&self, block: &CryptoHash) -> Result<Option<Data>, BlockTreeError> {
+        Ok(self.0.block_data(block)?)
+    }
+
+    pub(crate) fn block_datum(&self, block: &CryptoHash, datum_index: u32) -> Option<Datum> {
+        self.0.block_datum(block, datum_index)
+    }
+
+    pub(crate) fn block_at_height(&self, height: BlockHeight) -> Result<Option<CryptoHash>, BlockTreeError> {
+        Ok(self.0.block_at_height(height)?)
+    }
+
+    pub(crate) fn children(&self, block: &CryptoHash) -> Result<ChildrenList, BlockTreeError> {
+        Ok(self.0.children(block)?)
+    }
+
+    pub(crate) fn committed_app_state(&self, key: &[u8]) -> Option<Vec<u8>> {
+        self.0.committed_app_state(key)
+    }
+
+    pub(crate) fn pending_app_state_updates(&self, block: &CryptoHash) -> Result<Option<AppStateUpdates>, BlockTreeError> {
+        Ok(self.0.pending_app_state_updates(block)?)
+    }
+
+    pub(crate) fn committed_validator_set(&self) -> Result<ValidatorSet, BlockTreeError> {
+        Ok(self.0.committed_validator_set()?)
+    }
+
+    pub(crate) fn validator_set_updates_status(&self, block: &CryptoHash) -> Result<ValidatorSetUpdatesStatus, BlockTreeError> {
+        Ok(self.0.validator_set_updates_status(block)?)
+    }
+
+    pub(crate) fn locked_qc(&self) -> Result<QuorumCertificate, BlockTreeError> {
+        Ok(self.0.locked_qc()?)
+    }
+
+    pub(crate) fn highest_view_entered(&self) -> Result<ViewNumber, BlockTreeError> {
+        Ok(self.0.highest_view_entered()?)
+    }
+
+    pub(crate) fn highest_qc(&self) -> Result<QuorumCertificate, BlockTreeError> {
+        Ok(self.0.highest_qc()?)
+    }
+
+    pub(crate) fn highest_committed_block(&self) -> Result<Option<CryptoHash>, BlockTreeError> {
+        Ok(self.0.highest_committed_block()?)
+    }
+
+    pub(crate) fn newest_block(&self) -> Result<Option<CryptoHash>, BlockTreeError> {
+        Ok(self.0.newest_block()?)
+    }
+
+    pub(crate) fn highest_tc(&self) -> Result<Option<TimeoutCertificate>, BlockTreeError> {
+        Ok(self.0.highest_tc()?)
+    }
+
+    pub(crate) fn previous_validator_set(&self) -> Result<ValidatorSet, BlockTreeError> {
+        Ok(self.0.previous_validator_set()?)
+    }
+
+    pub(crate) fn validator_set_update_block_height(&self) -> Result<Option<BlockHeight>, BlockTreeError> {
+        Ok(self.0.validator_set_update_block_height()?)
+    }
+
+    pub(crate) fn validator_set_update_complete(&self) -> Result<bool, BlockTreeError> {
+        Ok(self.0.validator_set_update_complete()?)
+    }
+
+    pub(crate) fn validator_set_state(&self) -> Result<ValidatorSetState, BlockTreeError> {
+        Ok(self.0.validator_set_state()?)
+    }
+
+    pub(crate) fn highest_view_voted(&self) -> Result<Option<ViewNumber>, BlockTreeError> {
+        Ok(self.0.highest_view_voted()?)
+    }
+
+}
+
+#[derive(Debug)]
+pub enum BlockTreeError {
```

Comment:
General question: what is the rule that we use to choose between propagating an error upward vs panicking immediately? How far do we propagate upwards before we decide to panic? 

Based on what I'm seeing, we seem to propagate every kind of error upwards as far as possible, all the way to the user. Is this the rule?

Do we make a distinction between irrecoverable errors and recoverable errors?

Reply:
We propagate upwards up to the top-level method(s) of each thread, e.g., we panic from `Algorithm` thread's `execute`. 

Of course, we recover if the error is recoverable (though let me know if you see any instances where we don't?). But in our case very few errors are recoverable, most have to do with failed block tree access which is required for progress. 

My reasoning is this: this library is effectively an executable protocol where most relevant parameters as well as the underlying key value store can be configured by the user. Hence, when the user configures everything according our recommendations, these unrecoverable errors should not arise. However, with some strange, edge-case, parameter settings, these errors may arise. Hence, we may want to panic to inform the user early that something is wrong. This is also why the errors are propagated and wrapped the way they are: to provide a meaningful error trace. 

To make the error trace even more meaningful, I have been thinking of using [thiserror](https://docs.rs/thiserror/latest/thiserror/). Let me know what you think about this?

Reply:
I wanted to note a concrete problem with "propagate all errors upward and let the highest level decide whether to panic or not" that we discussed in yesterday's review meeting.

The first problem is that in order for the highest level to decide whether to panic or not, it needs to understand the error, but if we keep propagating errors upward, understanding the error entails understanding super deep levels of the library, which is infeasible and violates the principle of least knowledge.

I still hold that it's better to panic as soon as possible when we know that an error is caused by a bug. I don't like the idea of combining errors that are bugs with errors that are caused by temporary problems (e.g., too much traffic) or by mistakes in user input in the same error enums and having to disambiguate in the documentation that "this error is a bug, if this library was programmed correctly, then this variant will never be constructed").

Reply:
We only propagate errors to the highest thread level because they are irrecoverable errors that cannot be handled... So the highest level components does not need to understand the lower levels. And even if they were to be handled, that's still desirable. Consider an example of a company where a decision has to be made regarding some failure: if an employee does not know how to handle the failure, they ask their manager. Now if the manager know how to handle the failure, the handle it, else they ask the CEO to decide. So the CEO is like the top-level caller, the ultimate decision-maker.

In our case, all errors are of the same kind: errors that should not happen, yet may happen in some edge cases due to misconfiguration by the user, or DB malfunctioning. Basically, the source is always external to hotstuff-rs code, assuming our code is correct. But if it is not, then you can't distinguish between user's misconfiguration and hotstuff-rs error, can you?

I stand by the type-level error propagation, and I don't see any issues with it other than the fact that it requires more effort - but type-safe programming always requires more effort. It's the functional programming way of doing things, endorsed by many Rust programmers too. There are many good blog posts on it, such as [this one on error handling in Haskell](https://www.schoolofhaskell.com/school/starting-with-haskell/basics-of-haskell/10_Error_Handling). The decision is up to you!